Attached clusters

hankfreund opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
- Added new resource for attached clusters `google_container_attached_cluster`
- ~Added new datasource for retrieving valid attached cluster platform versions `google_container_attached_versions`~
- ~Added new datasource for retrieving a manifest to apply to the user cluster for boot-strapping `google_container_attached_install_manifest`~



<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_container_attached_cluster`
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I've detected that you're a community contributor. @melinath, a repository maintainer, has been assigned to assist you and help review your changes. 

<details>
  <summary>:question: First time contributing? Click here for more details</summary>

---

Your assigned reviewer will help review your code by: 
* Ensuring it's backwards compatible, covers common error cases, etc.
* Summarizing the change into a user-facing changelog note.
* Passes tests, either our "VCR" suite, a set of presubmit tests, or with manual test runs.

You can help make sure that review is quick by running local tests and ensuring they're passing in between each push you make to your PR's branch. Also, try to leave a comment with each push you make, as pushes generally don't generate emails.

If your reviewer doesn't get back to you within a week after your most recent change, please feel free to leave a comment on the issue asking them to take a look! In the absence of a dedicated review dashboard most maintainers manage their pending reviews through email, and those will sometimes get lost in their inbox.

---

</details>

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6858-old..auto-pr-6858) ( 12 files changed, 2267 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6858-old..auto-pr-6858) ( 12 files changed, 2267 insertions(+), 2 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6858-old..auto-pr-6858) ( 5 files changed, 400 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-6858-old..auto-pr-6858) ( 8 files changed, 263 insertions(+))

modular-magician commented:
#### Tests analytics 
Total tests: `2282` 
Passed tests `2033` 
Skipped tests: `244` 
Failed tests: `5` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccLoggingBucketConfigProject_cmekSettings|TestAccDataSourceGoogleContainerAttachedVersions|TestAccDataSourceGoogleContainerAttachedInstallManifest|TestAccContainerAttachedCluster_containerAttachedClusterFullExample|TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccLoggingBucketConfigProject_cmekSettings`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/recording/TestAccLoggingBucketConfigProject_cmekSettings.log)] 

Tests failed during RECORDING mode:
 `TestAccDataSourceGoogleContainerAttachedVersions`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/build-log/recording_build/TestAccDataSourceGoogleContainerAttachedVersions_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/recording/TestAccDataSourceGoogleContainerAttachedVersions.log)]
`TestAccDataSourceGoogleContainerAttachedInstallManifest`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/build-log/recording_build/TestAccDataSourceGoogleContainerAttachedInstallManifest_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/recording/TestAccDataSourceGoogleContainerAttachedInstallManifest.log)]
`TestAccContainerAttachedCluster_containerAttachedClusterFullExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/build-log/recording_build/TestAccContainerAttachedCluster_containerAttachedClusterFullExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/recording/TestAccContainerAttachedCluster_containerAttachedClusterFullExample.log)]
`TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/build-log/recording_build/TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/recording/TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/8239f10c-7425-47d4-a722-be9c96fad26e/recording) for each test

hankfreund commented:
The new acceptance tests will fail because the service isn't released yet. While developing, I tested against a test service. I was hoping to get the review out there and started so the code is as close to ready to merge as possible by the time we do release.

melinath reviewed (changes requested):
This is a fairly large PR - please split the data sources off into a separate PR.

melinath commented:
(Or make this a PR for the datasources and leave attached clusters for a separate PR.)

melinath commented:
oh, just saw your message re: test service. In general, we don't have time to review things that are not using working services.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6858-old..auto-pr-6858) ( 12 files changed, 2267 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6858-old..auto-pr-6858) ( 12 files changed, 2267 insertions(+), 2 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6858-old..auto-pr-6858) ( 5 files changed, 400 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-6858-old..auto-pr-6858) ( 8 files changed, 263 insertions(+))

modular-magician commented:
#### Tests analytics 
Total tests: `2306` 
Passed tests `2048` 
Skipped tests: `244` 
Failed tests: `14` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccDataFusionInstanceIamPolicyGenerated|TestAccDataFusionInstanceIamMemberGenerated|TestAccDataFusionInstanceIamBindingGenerated|TestAccDataFusionInstance_dataFusionInstanceEventExample|TestAccDataFusionInstance_dataFusionInstanceFullExample|TestAccDataFusionInstance_dataFusionInstanceZoneExample|TestAccContainerAttachedCluster_containerAttachedClusterFullExample|TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample|TestAccDataSourceGoogleContainerAttachedVersions|TestAccDataSourceGoogleContainerAttachedInstallManifest|TestAccVertexAIIndex_updated|TestAccVertexAIIndex_vertexAiIndexExample|TestAccLoggingBucketConfigProject_cmekSettings|TestAccVertexAIIndex_vertexAiIndexStreamingExample </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccDataFusionInstanceIamPolicyGenerated`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccDataFusionInstanceIamPolicyGenerated.log)]
`TestAccDataFusionInstanceIamMemberGenerated`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccDataFusionInstanceIamMemberGenerated.log)]
`TestAccDataFusionInstanceIamBindingGenerated`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccDataFusionInstanceIamBindingGenerated.log)]
`TestAccDataFusionInstance_dataFusionInstanceEventExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccDataFusionInstance_dataFusionInstanceEventExample.log)]
`TestAccDataFusionInstance_dataFusionInstanceFullExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccDataFusionInstance_dataFusionInstanceFullExample.log)]
`TestAccDataFusionInstance_dataFusionInstanceZoneExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccDataFusionInstance_dataFusionInstanceZoneExample.log)]
`TestAccVertexAIIndex_updated`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccVertexAIIndex_updated.log)]
`TestAccVertexAIIndex_vertexAiIndexExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccVertexAIIndex_vertexAiIndexExample.log)]
`TestAccLoggingBucketConfigProject_cmekSettings`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccLoggingBucketConfigProject_cmekSettings.log)]
`TestAccVertexAIIndex_vertexAiIndexStreamingExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccVertexAIIndex_vertexAiIndexStreamingExample.log)] 

Tests failed during RECORDING mode:
 `TestAccContainerAttachedCluster_containerAttachedClusterFullExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/build-log/recording_build/TestAccContainerAttachedCluster_containerAttachedClusterFullExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccContainerAttachedCluster_containerAttachedClusterFullExample.log)]
`TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/build-log/recording_build/TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample.log)]
`TestAccDataSourceGoogleContainerAttachedVersions`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/build-log/recording_build/TestAccDataSourceGoogleContainerAttachedVersions_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccDataSourceGoogleContainerAttachedVersions.log)]
`TestAccDataSourceGoogleContainerAttachedInstallManifest`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/build-log/recording_build/TestAccDataSourceGoogleContainerAttachedInstallManifest_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording/TestAccDataSourceGoogleContainerAttachedInstallManifest.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/51aeffe0-20b3-4e7b-94a2-30a9a47475b8/recording) for each test

melinath reviewed (commented):
removing from review queue

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6858-old..auto-pr-6858) ( 8 files changed, 1921 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6858-old..auto-pr-6858) ( 8 files changed, 1921 insertions(+), 2 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6858-old..auto-pr-6858) ( 5 files changed, 390 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-6858-old..auto-pr-6858) ( 8 files changed, 253 insertions(+))

modular-magician commented:
#### Tests analytics 
Total tests: `2355` 
Passed tests `2103` 
Skipped tests: `247` 
Failed tests: `5` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccContainerAttachedCluster_containerAttachedClusterFullExample|TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample|TestAccFirebaserulesRelease_BasicRelease|TestAccLoggingBucketConfigProject_cmekSettings|TestAccContainerCluster_withInvalidGatewayApiConfigChannel </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccContainerAttachedCluster_containerAttachedClusterFullExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/028b95e1-8033-4068-94f9-6c28ebbd235f/recording/TestAccContainerAttachedCluster_containerAttachedClusterFullExample.log)]
`TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/028b95e1-8033-4068-94f9-6c28ebbd235f/recording/TestAccContainerAttachedCluster_containerAttachedClusterMinimalExample.log)]
`TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/028b95e1-8033-4068-94f9-6c28ebbd235f/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccLoggingBucketConfigProject_cmekSettings`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/028b95e1-8033-4068-94f9-6c28ebbd235f/recording/TestAccLoggingBucketConfigProject_cmekSettings.log)]
`TestAccContainerCluster_withInvalidGatewayApiConfigChannel`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/028b95e1-8033-4068-94f9-6c28ebbd235f/recording/TestAccContainerCluster_withInvalidGatewayApiConfigChannel.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/028b95e1-8033-4068-94f9-6c28ebbd235f/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/028b95e1-8033-4068-94f9-6c28ebbd235f/recording) for each test

hankfreund commented:
Hi @melinath - I've split the PR and removed the data sources. I will send that PR out later. The acceptance tests now pass as we released the service.

melinath commented on mmv1/products/containerattached/terraform.yaml:
```diff
@@ -0,0 +1,64 @@
+overrides: !ruby/object:Overrides::ResourceOverrides
+  Cluster: !ruby/object:Overrides::Terraform::ResourceOverride
+    # Don't generate the LRO handling code because it doesn't account for regional services.
+    autogen_async: false
```
Rather than skipping autogeneration, you should be able to override the full URL for for the Cluster resource's LRO handling in api.yaml. For example: https://github.com/GoogleCloudPlatform/magic-modules/blob/c44fabcafc20de305a583fd27a3eb94df337e1ee/mmv1/products/dialogflowcx/api.yaml#L993

hankfreund replied:
I actually did try that. The generated code is:
`url := fmt.Sprintf("https://{{location}}-gkemulticloud.googleapis.com/v1/%s", w.CommonOperationWaiter.Op.Name)`

which is obviously problematic. You'll notice also that dialogflowcx does not auto-generate operation code. I couldn't find any regionalized service that does.

melinath replied:
yikes; opened https://github.com/hashicorp/terraform-provider-google/issues/13337 to track.

melinath commented on mmv1/products/containerattached/terraform.yaml:
```diff
@@ -0,0 +1,64 @@
+    import_format: ["projects/{{project}}/locations/{{location}}/attachedClusters/{{name}}"]
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "container_attached_cluster_minimal"
```
by convention we call this the "basic" test

```suggestion
        name: "container_attached_cluster_basic"
```

hankfreund replied:
Updated.

melinath commented on mmv1/products/containerattached/terraform.yaml:
```diff
@@ -0,0 +1,64 @@
+      # No diff suppression because it is still possible to explicitly clear the config by passing
+      # an empty value.
+      loggingConfig: !ruby/object:Overrides::Terraform::PropertyOverride
+        default_from_api: true
```
It might be worth double-checking that this behaves as expected; I'm not sure we have resources that use send_empty_value + allow_empty_object _and also_ default_from_api together yet.

hankfreund replied:
The basic test covers getting a default value from the service, the full test covers setting a value explicitly, and the (newly added) update test covers setting and then completely clearing the value. I think that's all the important scenarios? Is there something else I missed?

melinath replied:
hmm. so, thinking through it some more: the thing I'm concerned about is that this field seems to have the following potential values when you send it to the API:

1. Not set: API does nothing; values of this field are unmanaged by Terraform but will be stored in state.
2. Empty object: API clears field.
3. Non-empty object: API sets values according to object.

The weird thing here is that if you start with a non-empty object and then switch to "not set", the values for this field would (I think) be persisted in state indefinitely while not being present in the config.

Even though that's technically not an error, it seems like a pretty confusing / leaky user experience. I would expect that when I delete things from the config, they also get deleted in the API.

I think we should probably just not have this be a default_from_api field in Terraform - that is, make it required to send on every request. The problem is, that would require users to explicitly specify the field on every request - also not great. Maybe it could be marked optional but have a default value of an empty object that gets sent if the field is not set?

What do you think?

hankfreund replied:
On second thought, your original suggestion of sending an empty object when the user doesn't set anything does sound better. I'll do that.

melinath commented on mmv1/products/containerattached/terraform.yaml:
```diff
@@ -0,0 +1,64 @@
+      # If the user omits the monitoring config, the server will still respond with an empty config, which
+      # causes diff problems.
+      monitoringConfig: !ruby/object:Overrides::Terraform::PropertyOverride
+        diff_suppress_func: 'suppressAttachedClustersMonitoringConfigDiff'
```
It looks like this is similar in purpose to [`emptyOrUnsetBlockDiffSuppress`](https://github.com/hashicorp/terraform-provider-google/blob/6be5859718f35b1af72d4ca00a610e488ea3d854/google/common_diff_suppress.go#L113), which has a note:

> Using Optional + Computed is *strongly* preferred to this DSF, as it's more well understood and resilient to API changes.

Given that - default_from_api may be better here.

hankfreund replied:
Yeah, `default_from_api` seems to work and is less confusing.

melinath commented on mmv1/products/containerattached/api.yaml:
```diff
@@ -0,0 +1,273 @@
+          Configuration related to the cluster RBAC settings.
+        properties:
+          - !ruby/object:Api::Type::Array
+            name: admin_users
```
Fields in api.yaml use camelCase - it's probably not a big deal since this isn't a real API field but for consistency.

```suggestion
            name: adminUsers
```

Using this field name with custom expand & flatten is consistent with the aws and azure clusters.

melinath commented on mmv1/templates/terraform/custom_expand/containerattached_cluster_fleet_project.go.erb:
```diff
@@ -0,0 +1,22 @@
+	# See the License for the specific language governing permissions and
+	# limitations under the License.
+-%>
+func expand<%= prefix -%><%= titlelize_property(property) -%>(v interface{}, d TerraformResourceData, config *Config) (interface{}, error) {
```
This expander doesn't seem to match the behavior of the AWS & Azure clusters. Does the API reject requests that are not prepended with `projects/`?

hankfreund replied:
That is correct. I personally tend to forget about that requirement so I figured it'd be nice to have this additional help.

melinath replied:
this does seem convenient, but I'm concerned about then getting requests to add it on the AWS and Azure versions (which would be trickier.) alternatively, you could add a validation function like [StringMatch](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/validation#StringMatch) that ensures that the provided value starts with `projects/`... what do you think?

hankfreund replied:
I added a regex validation to check -- it seemed simplest.

melinath commented on mmv1/templates/terraform/examples/container_attached_cluster_full.tf.erb:
```diff
@@ -0,0 +1,34 @@
+      enabled = true
+    }
+  }
+  project = data.google_project.project.project_id
```
by convention, the project & location usually go at the top of the resource, since they're foundational to identifying the resource. not a blocker.

melinath commented on mmv1/templates/terraform/examples/container_attached_cluster_full.tf.erb:
```diff
@@ -0,0 +1,34 @@
+      project = data.google_project.project.number
+  }
+  logging_config {
+	  component_config {
```
```suggestion
    component_config {
```

melinath commented on mmv1/products/containerattached/api.yaml:
```diff
@@ -0,0 +1,273 @@
+                description: |
+                  The components of the logging configuration.
+                values:
+                  - :COMPONENT_UNSPECIFIED
```
Generally I prefer not to include UNSPECIFIED enum values, since they generally do nothing compared to not providing a value at all, and sometimes even cause errors. Just wanted to double-check that this is a meaningful value here?

hankfreund replied:
It is not meaningful. I have removed it.

melinath commented on mmv1/products/containerattached/api.yaml:
```diff
@@ -0,0 +1,273 @@
+        name: monitoringConfig
+        description: |
+          Monitoring configuration.
+        allow_empty_object: true
```
```suggestion
        allow_empty_object: true
        send_empty_value: true
```

melinath commented on mmv1/products/containerattached/api.yaml:
```diff
@@ -0,0 +1,273 @@
+            name: managedPrometheusConfig
+            description: |
+              Enable Google Cloud Managed Service for Prometheus in the cluster.
+            allow_empty_object: true
```
```suggestion
            allow_empty_object: true
            send_empty_value: true
```

melinath commented on mmv1/templates/terraform/examples/container_attached_cluster_full.tf.erb:
```diff
@@ -0,0 +1,34 @@
+      issuer_url = "https://oidc.issuer.url"
+      jwks = base64encode("{\"keys\":[{\"use\":\"sig\",\"kty\":\"RSA\",\"kid\":\"testid\",\"alg\":\"RS256\",\"n\":\"somedata\",\"e\":\"AQAB\"}]}")
+  }
+  platform_version = "1.24.0-gke.1"
```
This may already be the plan - once the datasource is available, it would be great to replace this with a reference to the datasource.

hankfreund replied:
That is the plan. One of the added datasources retrieves all valid versions from the service, so I will replace the hard-coded ones in the tests.

melinath reviewed (changes requested):
In addition to the requested changes below, please add an [update test](https://googlecloudplatform.github.io/magic-modules/docs/how-to/add-mmv1-test/#update-tests)

hankfreund commented on mmv1/products/containerattached/api.yaml:
```diff
@@ -0,0 +1,273 @@
+    update_mask: true
+    delete_verb: :DELETE
+    description: |
+      An Anthos attached cluster.
```
Yeah, I think "customer owned" is more correct.

hankfreund commented:
Made all requested changes including the update test. Uncovered an update mask bug. Thanks!

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6858-old..auto-pr-6858) ( 9 files changed, 2052 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6858-old..auto-pr-6858) ( 9 files changed, 2052 insertions(+), 2 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6858-old..auto-pr-6858) ( 5 files changed, 386 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-6858-old..auto-pr-6858) ( 8 files changed, 253 insertions(+))

modular-magician commented:
#### Tests analytics 
Total tests: `2376` 
Passed tests `2122` 
Skipped tests: `248` 
Failed tests: `6` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccLoggingBucketConfigProject_cmekSettings|TestAccFirebaserulesRelease_BasicRelease|TestAccContainerCluster_withInvalidGatewayApiConfigChannel|TestAccContainerAttachedCluster_update|TestAccContainerAttachedCluster_containerAttachedClusterBasicExample|TestAccComputeForwardingRule_update </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccLoggingBucketConfigProject_cmekSettings`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/003be1b8-218b-4b36-bb9b-47306786e8ba/recording/TestAccLoggingBucketConfigProject_cmekSettings.log)]
`TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/003be1b8-218b-4b36-bb9b-47306786e8ba/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccContainerCluster_withInvalidGatewayApiConfigChannel`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/003be1b8-218b-4b36-bb9b-47306786e8ba/recording/TestAccContainerCluster_withInvalidGatewayApiConfigChannel.log)]
`TestAccContainerAttachedCluster_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/003be1b8-218b-4b36-bb9b-47306786e8ba/recording/TestAccContainerAttachedCluster_update.log)]
`TestAccContainerAttachedCluster_containerAttachedClusterBasicExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/003be1b8-218b-4b36-bb9b-47306786e8ba/recording/TestAccContainerAttachedCluster_containerAttachedClusterBasicExample.log)]
`TestAccComputeForwardingRule_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/003be1b8-218b-4b36-bb9b-47306786e8ba/recording/TestAccComputeForwardingRule_update.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/003be1b8-218b-4b36-bb9b-47306786e8ba/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/003be1b8-218b-4b36-bb9b-47306786e8ba/recording) for each test

melinath commented on mmv1/products/containerattached/api.yaml:
```diff
@@ -0,0 +1,273 @@
+                  - :SYSTEM_COMPONENTS
+                  - :WORKLOADS
+      - !ruby/object:Api::Type::NestedObject
+        name: errors
```
I think this is inverted - it should be an array of objects where each object has a message key. Currently it would give output like this instead:

```
errors {
    message = [
        "error 1",
        "error 2",
        "error 3",
    ]
}
```

hankfreund replied:
Yeah, that makes sense!

melinath commented on mmv1/products/containerattached/api.yaml:
```diff
@@ -0,0 +1,273 @@
+        name: monitoringConfig
+        description: |
+          Monitoring configuration.
+        allow_empty_object: true
```
Just want to double-check (since monitoringConfig and managedPrometheusConfig are both optional) that these all actually need to be allow_empty_object & send_empty_value? Although we can support that it's generally something we prefer not to do unless it adds value over just not sending the fields at all.

melinath commented on mmv1/products/containerattached/api.yaml:
```diff
@@ -0,0 +1,273 @@
+
+--- !ruby/object:Api::Product
+name: ContainerAttached
+display_name: Anthos Attached clusters
```
It might make sense to leave this as ContainerAttached to match the other resources in the provider registry

melinath commented on mmv1/third_party/terraform/tests/resource_container_attached_cluster_update_test.go:
```diff
@@ -0,0 +1,113 @@
+`, context)
+}
+
+func testAccContainerAttachedCluster_containerAttachedCluster_update(context map[string]interface{}) string {
```
since this is a complex resource it could be worth using `lifecycle.prevent_destroy` to make sure no input-only fields are accidentally modified & triggering recreation. (this is currently documented at https://github.com/hashicorp/terraform-provider-google/issues/12594)

hankfreund replied:
Good call! Updated.

melinath reviewed (changes requested):
A few changes noted below. Also, resource_container_attached_cluster_update_test will need to have gofmt run on it - currently it's causing the provider unit tests to fail lint checks.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6858-old..auto-pr-6858) ( 9 files changed, 2052 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6858-old..auto-pr-6858) ( 9 files changed, 2052 insertions(+), 2 deletions(-))
TF Validator: [Diff](https://github.com/modular-magician/terraform-validator/compare/auto-pr-6858-old..auto-pr-6858) ( 5 files changed, 386 insertions(+), 3 deletions(-))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-6858-old..auto-pr-6858) ( 8 files changed, 253 insertions(+))

modular-magician commented:
#### Tests analytics 
Total tests: `2382` 
Passed tests `2133` 
Skipped tests: `248` 
Failed tests: `1` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccContainerCluster_withInvalidGatewayApiConfigChannel </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccContainerCluster_withInvalidGatewayApiConfigChannel`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/330fdcc6-a0ba-4936-b565-997ff71d5b68/recording/TestAccContainerCluster_withInvalidGatewayApiConfigChannel.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/330fdcc6-a0ba-4936-b565-997ff71d5b68/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/330fdcc6-a0ba-4936-b565-997ff71d5b68/recording) for each test

hankfreund commented:
Made the requested changes and formatted the update test.

modular-magician commented:
#### Tests analytics 
Total tests: `2382` 
Passed tests `2130` 
Skipped tests: `248` 
Failed tests: `4` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccFirebaserulesRelease_BasicRelease|TestAccContainerCluster_withInvalidGatewayApiConfigChannel|TestAccContainerAttachedCluster_update|TestAccContainerAttachedCluster_containerAttachedClusterBasicExample </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/7f6c2771-7c5e-4b98-b38a-70249371907d/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccContainerCluster_withInvalidGatewayApiConfigChannel`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/7f6c2771-7c5e-4b98-b38a-70249371907d/recording/TestAccContainerCluster_withInvalidGatewayApiConfigChannel.log)]
`TestAccContainerAttachedCluster_update`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/7f6c2771-7c5e-4b98-b38a-70249371907d/recording/TestAccContainerAttachedCluster_update.log)]
`TestAccContainerAttachedCluster_containerAttachedClusterBasicExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/7f6c2771-7c5e-4b98-b38a-70249371907d/recording/TestAccContainerAttachedCluster_containerAttachedClusterBasicExample.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/7f6c2771-7c5e-4b98-b38a-70249371907d/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6858/artifacts/7f6c2771-7c5e-4b98-b38a-70249371907d/recording) for each test
